### ISSUE TITLE
Double Login Required After Restart

### ISSUE THREAD

**Issue by dirkniblick** [rundeck/rundeck#2702]
Issue type: Bug report

**My Rundeck detail**

* Rundeck version: 2.9.2-1
* install type: rpm
* OS Name/version: CentOS Linux release 7.2.1511
* DB Type/version: h2

**Expected Behavior**
Should only need to login once.

**Actual Behavior**
After starting or restarting rundeckd, everyone authenticating via LDAP has to authenticate *twice* before they're actually let in.  They don't receive an error message.  The login page just shows up again, but on the second time it'll work.  You can see below the attempt at 08:03:37 seems to work okay, but I have to try again 12 seconds later with the only difference being a hit in the cache.

```
2017-08-18 08:03:37.316:INFO:cdrjj.JettyCachingLdapLoginModule:qtp1480010240-19: Login attempts: 1, Hits: 0, Ratio: 0%.
2017-08-18 08:03:37.316:DBUG:cdrjj.JettyCachingLdapLoginModule:qtp1480010240-19: Cache Miss for bobby.
2017-08-18 08:03:37.317:DBUG:cdrjj.JettyCachingLdapLoginModule:qtp1480010240-19: Searching for users with filter: '(&(objectClass={0})({1}={2}))' from base dn: DC=xyz,DC=xyz,DC=com
2017-08-18 08:03:37.320:DBUG:cdrjj.JettyCachingLdapLoginModule:qtp1480010240-19: Found user?: true
2017-08-18 08:03:37.320:INFO:cdrjj.JettyCachingLdapLoginModule:qtp1480010240-19: Attempting authentication: CN=Bobby Bouchet,OU=Users,OU=Information Technology,DC=xyz,DC=xyz,DC=com
2017-08-18 08:03:37.795:DBUG:cdrjj.JettyCachingLdapLoginModule:qtp1480010240-19: JettyCachingLdapLoginModule: User 'bobby' has roles: [Programming, Folder Redirection, RunDeckUsers, RunDeckAdministrators, Administrators, Abuse, remalert, loialert, newFormula, root, ticket_notifications, readyToFlip, programmingandoffice]
2017-08-18 08:03:37.796:DBUG:cdrjj.JettyCachingLdapLoginModule:qtp1480010240-19: Adding bobby set to expire: 150306141779630000
2017-08-18 08:03:49.326:INFO:cdrjj.JettyCachingLdapLoginModule:qtp1480010240-1947: Login attempts: 2, Hits: 0, Ratio: 0%.
2017-08-18 08:03:49.326:DBUG:cdrjj.JettyCachingLdapLoginModule:qtp1480010240-1947: Cache Hit for bobby.
```

I've disabled caching (cacheDurationMillis="0"), but the same thing happens, just without caching the LDAP information:

```
2017-08-18 08:14:00.645:INFO:cdrjj.JettyCachingLdapLoginModule:qtp1480010240-15: Login attempts: 1, Hits: 0, Ratio: 0%.
2017-08-18 08:14:00.646:DBUG:cdrjj.JettyCachingLdapLoginModule:qtp1480010240-15: Searching for users with filter: '(&(objectClass={0})({1}={2}))' from base dn: DC=xyz,DC=xyz,DC=com
2017-08-18 08:14:00.651:DBUG:cdrjj.JettyCachingLdapLoginModule:qtp1480010240-15: Found user?: true
2017-08-18 08:14:00.651:INFO:cdrjj.JettyCachingLdapLoginModule:qtp1480010240-15: Attempting authentication: CN=Bobby Bouchet,OU=Users,OU=Information Technology,DC=xyz,DC=xyz,DC=com
2017-08-18 08:14:00.907:DBUG:cdrjj.JettyCachingLdapLoginModule:qtp1480010240-15: JettyCachingLdapLoginModule: User 'bobby' has roles: [Programming, Folder Redirection, RunDeckUsers, RunDeckAdministrators, Administrators, Abuse, remalert, loialert, newFormula, root, ticket_notifications, readyToFlip, programmingandoffice]
2017-08-18 08:14:05.510:INFO:cdrjj.JettyCachingLdapLoginModule:qtp1480010240-15: Login attempts: 2, Hits: 0, Ratio: 0%.
2017-08-18 08:14:05.511:DBUG:cdrjj.JettyCachingLdapLoginModule:qtp1480010240-15: Searching for users with filter: '(&(objectClass={0})({1}={2}))' from base dn: DC=xyz,DC=xyz,DC=com
2017-08-18 08:14:05.512:DBUG:cdrjj.JettyCachingLdapLoginModule:qtp1480010240-15: Found user?: true
2017-08-18 08:14:05.512:INFO:cdrjj.JettyCachingLdapLoginModule:qtp1480010240-15: Attempting authentication: CN=Bobby Bouchet,OU=Users,OU=Information Technology,DC=xyz,DC=xyz,DC=com
2017-08-18 08:14:05.798:DBUG:cdrjj.JettyCachingLdapLoginModule:qtp1480010240-15: JettyCachingLdapLoginModule: User 'bobby' has roles: [Programming, Folder Redirection, RunDeckUsers, RunDeckAdministrators, Administrators, Abuse, remalert, loialert, newFormula, root, ticket_notifications, readyToFlip, programmingandoffice]
```

This behavior exhibited itself before we upgraded from 2.8.4-1.  After that first double-login, the same will be able to login later in one try, at least until the service is restarted again.  I've been just ignoring it, but now more people are gaining access so the complaining is getting louder especially since I've implemented a daily [backup of the database and log files](http://rundeck.org/docs/administration/backup-and-recovery.html).

**How to reproduce Behavior**

This is my login.config file.  This behavior exhibits itself with an LDAP only config too (i.e., without the PropertyFileLoginModule section).

```
multipleauth {

  org.eclipse.jetty.jaas.spi.PropertyFileLoginModule sufficient
    debug="true"
    file="/etc/rundeck/realm.properties"
    refreshInterval="60"
    caseInsensitive="true";

  com.dtolabs.rundeck.jetty.jaas.JettyCachingLdapLoginModule required
    debug="true"
    contextFactory="com.sun.jndi.ldap.LdapCtxFactory"
    providerUrl="ldap://xxx.xxx.xxx.xxx:389"
    bindDn="CN=xyz,CN=xyz,DC=xyz,DC=xyz,DC=com"
    bindPassword="PASSWORD"
    authenticationMethod="simple"
    forceBindingLogin="true"
    userBaseDn="DC=xyz,DC=xyz,DC=com"
    userRdnAttribute="sAMAccountName"
    userIdAttribute="sAMAccountName"
    userPasswordAttribute="unicodePwd"
    userObjectClass="user"
    roleBaseDn="DC=xyz,DC=xyz,DC=com"
    roleNameAttribute="cn"
    roleMemberAttribute="member"
    roleObjectClass="group"
    cacheDurationMillis="30000"
    reportStatistics="true"
    timeoutRead="10000"
    timeoutConnect="20000"
    nestedGroups="true"
    storePass="true";
};
```

All the other config is pretty much unchanged from what the RPM provided, however I did switch out "user" for "RunDeckUsers" in the web.xml file (did the same thing before I did that).

**Comment by ahonor** [comment]
It's possible the grails.serverURL does not match the URL the user is putting into their browser. Can you check that?

**Comment by dirkniblick** [comment]
Yeah, I checked that before I opened the issue, but I figured I'd be sure.  I copied and pasted the URL from the file into my browser after restarting, but I decided on a whim to put it into MS Edge this time.  I got in on the first try.  I restarted and pasted it into Chrome...double login.  This time I actually looked at the URL and noticed this"/user/login;jsessionid=..." garbage.  I restarted again, pasted the URL again (i.e., http://hostname.com:4440) into the browser, hit enter, and either Rundeck or Chrome slid the jsessionid in there.  After the first login attempt, the nonsense after jsessionid changed and I was able to get in fine.    Other attempts with Edge always worked on the first try but I noticed it was adding the jsessionid to the URL too.

A minimal amount of Googling got me [this](https://stackoverflow.com/questions/8880592/chrome-and-jsessionid).  I know it's 5 years old, but it seems to look like the same thing.  It has a lot of the same vocabulary (e.g., grails, chrome, jsessionid, "login again").
